Shield modification warheads & minor fixes/additions

Starkku opened the pull request:
```
[SOMESHIELDTYPE]                     ; ShieldType name
InitialStrength=0                    ; integer
BreakWeapon=                         ; WeaponType

[SOMEWARHEAD]                        ; WarheadType
Shield.Penetrate=false               ; boolean
Shield.Break=false                   ; boolean
Shield.BreakAnim=                    ; animation
Shield.HitAnim=                      ; animation
Shhield.BreakWeapon=                 ; WeaponType
Shield.AbsorbPercent=                ; double, percents
Shield.PassPercent=                  ; double, percents
Shield.Respawn.Duration=0            ; integer, game frames
Shield.Respawn.Amount=0.0            ; double, percents or absolute
Shield.Respawn.Rate=-1.0             ; double, ingame minutes
Shield.Respawn.ResetTimer=false      ; double, ingame minutes
Shield.SelfHealing.Duration=0        ; integer, game frames
Shield.SelfHealing.Amount=0.0        ; double, percents or absolute
Shield.SelfHealing.Rate=-1.0         ; double, ingame minutes
Shield.SelfHealing .ResetTimer=false ; double, ingame minutes
Shield.AffectTypes=                  ; List of ShieldType names
Shield.AttachTypes=                  ; List of ShieldType names
Shield.RemoveTypes=                  ; List of ShieldType names
Shield.ReplaceOnly=false             ; boolean
Shield.ReplaceNonRespawning=false    ; boolean
Shield.InheritStateOnReplace=false   ; boolean
```

- `InitialStrength` can be used to set a different initial strength value from maximum.
- `BreakWeapon`, if set, will be fired at the TechnoType once the shield breaks.
- `Shield.BreakAnim` will be displayed instead of ShieldType `BreakAnim` if the shield is broken by the Warhead, either through damage or `Shield.Break`.
- `Shield.HitAnim` will be displayed instead of ShieldType `HitAnim` if set when Warhead hits the shield.
- `Shield.BreakWeapon` will be fired instead of ShieldType `BreakWeapon` if the shield is broken by the Warhead, either through damage or `Shield.Break`.
- `Shield.AbsorbPercent` overrides the `AbsorbPercent` value set in the ShieldType that is being damaged.
- `Shield.PassPercentShield` overrides the `PassPercent` value set in the ShieldType that is being damaged.
- `Shield.Respawn.Rate` & `Shield.Respawn.Amount` override ShieldType `Respawn.Rate` and `Respawn.Amount` for duration of `Shield.Respawn.Duration` amount of frames. Negative rate & zero or lower amount default to ShieldType values. If `Shield.Respawn.ResetTimer` is set, currently running shield respawn timer is reset, otherwise the timer's duration is adjusted to match `Shield.Respawn.Rate` without restarting the timer.  If the effect expires while respawn timer is running, remaining time is adjusted to match ShieldType `Respawn.Rate`. Re-applying the effect resets the duration to `Shield.Respawn.Duration`
- `Shield.SelfHealing.Rate` & `Shield.SelfHealing.Amount` override ShieldType `SelfHealing.Rate` and `SelfHealing.Amount` for duration of `Shield.SelfHealing.Duration` amount of frames. Negative rate & zero or lower amount default to ShieldType values. If `Shield.SelfHealing.ResetTimer` is set, currently running self-healing timer is restarted, otherwise timer's duration 'is adjusted to match `Shield.SelfHealing.Rate` without restarting the timer. If the effect expires while self-healing timer is running, remaining time is adjusted to match ShieldType `SelfHealing.Rate`. Re-applying the effect resets the duration to `Shield.SelfHealing.Duration`.
- `Shield.AffectsTypes` allows listing which ShieldTypes can be affected by any of the effects listed above. If none are listed, all ShieldTypes are affected.
- `Shield.AttachTypes` & `Shield.RemoveTypes` allows listing ShieldTypes that are attached or removed, respectively from any targets affected by the warhead (positive `Verses` values). Normally only first listed ShieldType in `Shield.AttachTypes` is applied.
  - If `Shield.ReplaceOnly` is set, shields from `Shield.AttachTypes` are only applied to affected targets from which shields were simultaneously removed, matching the order listed in `Shield.RemoveTypes`. If `Shield.AttachTypes` contains less items than `Shield.RemoveTypes`, last item from the former is used for any remaining removed shields.
   - If `Shield.ReplaceNonRespawning` is set, shield from `Shield.AttachTypes` replaces existing shields that have been broken and cannot respawn on their own.
   - If `Shield.InheritStateOnReplace` is set, shields replaced via `Shield.ReplaceOnly` inherit the current strength (relative to ShieldType `Strength`) of the previous shield and whether or not the shield was currently broken. Self-healing and respawn timers are always reset.

## Note the spelling changes for `PenetratesShield`, `BreaksShield`, `AbsorbPercentShield` and `PassPercentShield`!

In addition, following Phobos bugs were fixed:

- Shields being able to be damaged when under effects of `Temporal` warheads.
- `AbsorbPercentShield` and `PassPercentShield` not being saved properly.

Starkku commented:
Misc. additions & fixes.

- Implemented `BreakWeapon` & `Shield.BreakWeapon`, a weapon that is fired on the TechnoType with the shield when the shield breaks.
- Adjusted how `Shield.SelfHealing` & `Shield.Respawn` affect timers and added `ResetTimer` keys to control whether or not timers are reset on application.
- Made `Shield.BreakAnim` apply if Warhead breaks shield through damage.
- Implemented `Shield.HitAnim`.

Starkku commented:
Further adjustments to the timer logic.

- Added `Shield.Respawn.Duration` and made it run on timer instead of always applying to next shield respawn.
- Duration of respawn & self-healing warhead effects should now correctly reset on re-applying the effect.

YRIDZE commented on src/New/Entity/ShieldClass.cpp:
```diff
@@ -87,7 +99,7 @@ int ShieldClass::ReceiveDamage(args_ReceiveDamage* args)
 	const auto pWHExt = WarheadTypeExt::ExtMap.Find(args->WH);
 
-	if (!this->HP || *args->Damage == 0 || this->Techno->IsIronCurtained() || pWHExt->PenetratesShield)
+	if (this->Temporal || !this->HP || *args->Damage == 0 || this->Techno->IsIronCurtained() || pWHExt->Shield_Penetrate)
```
Please, split it up into a few lines.

YRIDZE reviewed (changes requested):
Code seems good. I found several problems in "if" statements. Please fix it.

Starkku commented:
- Fixed a potential issue with shield attaching, not sure what kind of problems it may have caused (it replaced the currently used ShieldType even if it couldn't replace shield (new type has Strength <=0 or the unit already had a shield).
- Added an option to allow replacing non-respawning shields via attaching a new one.

Metadorius commented on src/Ext/WarheadType/Body.h:
```diff
@@ -66,10 +91,30 @@ class WarheadTypeExt
+			, Shield_Respawn_Rate(-1)
+			, Shield_Respawn_Rate_InMinutes(-1.0)
+			, Shield_Respawn_ResetTimer(false)
+			, Shield_SelfHealing_Duration(0)
+		    , Shield_SelfHealing_Amount(0.0)
+		    , Shield_SelfHealing_Rate(-1)
+			, Shield_SelfHealing_Rate_InMinutes(-1.0)
```
```suggestion
			, Shield_SelfHealing_Duration(0)
			, Shield_SelfHealing_Amount(0.0)
			, Shield_SelfHealing_Rate(-1)
			, Shield_SelfHealing_Rate_InMinutes(-1.0)
```
